### DISCUSSION TITLE
Can we use ASTRA toolbox on acoustic tomography with special geometry?

### DISCUSSION THREAD

**Discussion by dannychancc** [astra-toolbox/astra-toolbox#371] [Q&A] [answered]
Hello everyone,

Recently I found out ASTRA toolbox for build tomography in Python.

However, for the acoustic tomography, the geometry is a little bit different. we put acoustic transceivers around a solid sample to find out what is its internal conditions. Each transceiver will send out sound waves, the rest will receive, distance between the transceivers may NOT be the same. The velocity of each path will be calculated in a matrix, then generate a tomogram of velocity distribution.  (Please see attachment below)

Is it possible to use ASTRA toolbox in this case, especially how to fit in to the projection geometry, not parallel, beam fan?
Is there any suggestion?

Thanks!

Danny

[Acoustic tomography.pdf](https://github.com/astra-toolbox/astra-toolbox/files/11120660/Acoustic.tomography.pdf)

**Comment by askorikov** [comment]
Hi Danny! I guess it should be technically possible to achieve this by using a 1 pixel wide detector with [`fanflat_vec`](https://www.astra-toolbox.com/docs/geom2d.html) geometry, where you can encode the positions of your transceivers and receivers. However, performance will probably not be great, because Astra is designed to efficiently propagate large bundles of rays in similar directions.

**Reply by dannychancc** [reply]
Hi Askorikov,

Thanks for your help! May I know whether I configure the source and detector position in every single ray? Suppose I simplify there are 4 Transceivers, so there are altogether 4 x 3 = 12 rays, so I can do like this?

proj_geom = astra_create_proj_gem('fanflat_vec', 12, vector);

vector will have 12 elements of :
vector = "{srcTx1_X, srcTx1_Y, dRx2_X, dRx2_Y, uX, uY},
                {srcTx1_X, srcTx1_Y, dRx3_X, dRx3_Y, uX, uY},
                {srcTx1_X, srcTx1_Y, dRx4_X, dRx4_Y, uX, uY},
                 .............
                 .............
                {srcTx4_X, srcTx4_Y, dRx1_X, dRx1_Y, uX, uY},
                {srcTx4_X, srcTx4_Y, dRx2_X, dRx2_Y, uX, uY},
                {srcTx4_X, srcTx4_Y, dRx3_X, dRx3_Y, uX, uY}"

Am I correct? Thanks a lot!
Danny

**Reply by askorikov** [reply]
Yes exactly, that's what I had in mind. Let us know if it doesn't work out.

**Reply by dannychancc** [reply]
Hi Askorikov,

Thanks for your help!

I've written the code as below, the number of transceivers is 8, so there are 56 paths. The document is a little bit simple so I have the following questions, can you help?

1. In the last 2 fanflat_vec projection geometry, the last 2 of the 6 elements in the vector should be filled with "u : the vector between the centers of detector pixels 0 and 1", what essentially does it mean?

2. What should I input in the det_count in astra.create_proj_geom('fanflat_vec', 1, vectors), it should be 1 or 56?

3. I have compile error on creating sinogram astra.data2d.create('-sino', proj_geom, S), the error is ValueError: The dimensions of the data do not match those specified in the geometry: (56,) != (56, 1). So what should be the projection data array has to be input (S: there is 56 paths, so there is 56 projection data)?

Thanks!!

```
import astra
import numpy as np
import math
import pylab

# create vol geometry
vol_geom = astra.create_vol_geom(100, 100)

# number of transceivers = 8
num_tx = 8

# number of path = 8*7 = 56
num_path = num_tx * (num_tx - 1)

# angle of the 8 transceivers, 8 equal spaced on circumference of a circle for testing
arr = [0, 45, 90, 135, 180, 225, 270, 315]

# distance from origin of the transceivers is 20
len = 20

#create projection geometry
vectors = np.zeros((num_path, 6))

for i in range(num_tx):

    for j in range(num_tx-1):

        vectors[i+j,0] = math.sin(math.radians(arr[i])) * len
        vectors[i+j,1] = -math.cos(math.radians(arr[i])) * len
        vectors[i+j,2] = math.sin(math.radians(arr[(i+j+1) % num_tx])) * len
        vectors[i+j,3] = -math.cos(math.radians(arr[(i+j+1) % num_tx])) * len
        vectors[i+j,4] = np.cos(math.radians(arr[(i+j+1) % num_tx])) * 1.0
        vectors[i+j,5] = np.sin(math.radians(arr[(i+j+1) % num_tx])) * 1.0

# the number of detectors should fill 1 or 56??
proj_geom = astra.create_proj_geom('fanflat_vec', 1, vectors)

# create projector
proj_id = astra.create_projector('line_fanflat', proj_geom, vol_geom)

# create sinogram, size = num_path = 56
S = np.array( [2007, 1808, 1661, 1428, 1631, 1867, 1922, 
     1755, 1807, 1427, 1393, 1710, 1828, 2007,
     1957, 1677, 1513, 1420, 1646, 1808, 1755,
     1753, 1719, 1616, 1470, 1661, 1807, 1957,
     1883, 1649, 1625, 1428, 1427, 1677, 1753,
     1735, 1719, 1631, 1393, 1513, 1719, 1883,
     1899, 1867, 1710, 1420, 1616, 1649, 1735,
     1922, 1828, 1646, 1470, 1625, 1719, 1899])

# Below: ValueError: The dimensions of the data do not match those specified in the geometry: (56,) != (56, 1)
sino_id = astra.data2d.create('-sino', proj_geom, S)

rec_id = astra.data2d.create('-vol', vol_geom)

# Configure algorithm
cfg = astra.astra_dict('SIRT')
cfg['ReconstructionDataId'] = rec_id
cfg['ProjectionDataId'] = sino_id
cfg['ProjectorDataId'] = proj_id
alg_id = astra.algorithm.create(cfg)

# Run
astra.algorithm.run(alg_id, 100)
rec = astra.data2d.get(rec_id)

pylab.figure(1)
pylab.imshow(rec)
pylab.show()

# Clean up. Note that GPU memory is tied up in the algorithm object,
# and main RAM in the data objects.
astra.algorithm.delete(alg_id)
astra.data2d.delete(rec_id)
astra.data2d.delete(sino_id)
astra.projector.delete(proj_id)

```

Danny

**Reply by askorikov** [reply]
1. It basically encodes the rotations and scaling of the detector, which in your case should be perpendicular to the incoming ray and have a length of 1 (will require some trigonometry to set up). 
2. It should be 1, because it's the number of pixels for a given direction.
3. You need to add a dimension to your `S` array (the first dimension is the number of pixels for a given direction, which is 1 in your case).

A couple more comments:
* There's a typo on line 63, should be `ProjectorId`
* Your volume geometry is now defined to be of size 100x100, which is much large than the `len`. Not sure if it's intended, and you can tune the size of the geometry by setting it like [`vol_geom = astra.create_vol_geom(rows, cols, min_x, max_x, min_y, max_y)`](https://www.astra-toolbox.com/_modules/astra/creators.html#create_vol_geom).
* The nested `for` loops don't fill the `vectors` array correctly, it should be `i*(num_tx-1) + j` in the index I guess.

**Reply by dannychancc** [reply]
Hi Askorikov,

Thanks for your comments!
I modified based on your comments, I am able to compile and output an image. (The code and the image are attached below)

```
import astra
import numpy as np
import math
import matplotlib.pyplot as plt

np.set_printoptions(threshold=np.inf)

# create vol geometry
vol_geom = astra.create_vol_geom(40, 40)

# number of transceivers = 8
num_tx = 8

# number of path = 8*7 = 56
num_path = num_tx * (num_tx - 1)

# angle of the 8 transceivers, 8 equal spaced on circumference of a circle for testing
arr = [0, 45, 90, 135, 180, 225, 270, 315]

# distance from origin of the transceivers is 20
len = 20

#create projection geometry
vectors = np.zeros((num_path, 6))


for i in range(num_tx):

    for j in range(num_tx-1):

        #print ("index: ", i, j, src_idx)
        vectors[i*(num_tx-1)+j][0] = math.sin(math.radians(arr[i])) * len
        vectors[i*(num_tx-1)+j][1] = -math.cos(math.radians(arr[i])) * len
        vectors[i*(num_tx-1)+j][2] = math.sin(math.radians(arr[(i+j+1) % num_tx])) * len
        vectors[i*(num_tx-1)+j][3] = -math.cos(math.radians(arr[(i+j+1) % num_tx])) * len
        vectors[i*(num_tx-1)+j][4] = np.cos(math.radians(arr[(i+j+1) % num_tx])) * 1.0
        vectors[i*(num_tx-1)+j][5] = np.sin(math.radians(arr[(i+j+1) % num_tx])) * 1.0

# the number of detectors should fill 1 or 56??
proj_geom = astra.create_proj_geom('fanflat_vec', 1, vectors)

# create projector
proj_id = astra.create_projector('line_fanflat', proj_geom, vol_geom)

# create sinogram, size = num_path = 56
#S = np.array( [[2007], [1808], [1661], [1428], [1631], [1867], [1922], 
#               [1755], [1807], [1427], [1393], [1710], [1828], [2007], 
#               [1957], [1677], [1513], [1420], [1646], [1808], [1755], 
#               [1753], [1719], [1616], [1470], [1661], [1807], [1957], 
#               [1883], [1649], [1625], [1428], [1427], [1677], [1753], 
#               [1735], [1719], [1631], [1393], [1513], [1719], [1883], 
#               [1899], [1867], [1710], [1420], [1616], [1649], [1735], 
#               [1922], [1828], [1646], [1470], [1625], [1719], [1899]])

S = np.array( [[2000], [1700], [1500], [1300], [1500], [1700], [2000], 
               [2000], [1700], [1500], [1300], [1500], [1700], [2000],
               [2000], [1700], [1500], [1300], [1500], [1700], [2000],
               [2000], [1700], [1500], [1300], [1500], [1700], [2000],
               [2000], [1700], [1500], [1300], [1500], [1700], [2000],
               [2000], [1700], [1500], [1300], [1500], [1700], [2000],
               [2000], [1700], [1500], [1300], [1500], [1700], [2000],
               [2000], [1700], [1500], [1300], [1500], [1700], [2000]])

sino_id = astra.data2d.create('-sino', proj_geom, S)

rec_id = astra.data2d.create('-vol', vol_geom)

# Configure algorithm
cfg = astra.astra_dict('SIRT')
cfg['ReconstructionDataId'] = rec_id
cfg['ProjectionDataId'] = sino_id
cfg['ProjectorId'] = proj_id
alg_id = astra.algorithm.create(cfg)

# Run
astra.algorithm.run(alg_id, 100)
rec = astra.data2d.get(rec_id)

fig = plt.figure
plt.imshow(rec)
plt.colorbar(orientation='vertical')
plt.show()

# Clean up. Note that GPU memory is tied up in the algorithm object,
# and main RAM in the data objects.
astra.algorithm.delete(alg_id)
astra.data2d.delete(rec_id)
astra.data2d.delete(sino_id)
astra.projector.delete(proj_id)

```

![Figure_1](https://user-images.githubusercontent.com/129504616/230329770-206d963c-6378-4bc4-acf7-90fa0fdb84e0.png)


The projection data is emulated as there is a hole inside the center, so the speed is becoming lower when the path is near center, so I expect the tomogram would be similar to the right hand side picture below:

![target](https://user-images.githubusercontent.com/129504616/230331439-c426fc8e-dbf2-4c26-8a2d-a22f9f7b381b.png)

So I want to know:
1. Is there any ways for the toolbox to produce similar picture like that? 
2. Or how to modify, does the toolbox have source code for us to investigate?

Danny

**Reply by dannychancc** [reply]
Hi Askorikov,

Do you have some comments?

Danny

**Reply by askorikov** [reply, marked as answer]
Hi! Honestly, I'm not quite sure, but it is probably the case that Astra toolbox is not suited for the task. It works with a tomography model where radiation is assumed to be propagating along straight lines, which is suited for "hard" radiation like X-rays. In your case (and in acoustic tomography in general as far as I'm aware), you probably need to work with full wave propagation, which goes beyond the scope of the toolbox. That being said, maybe you can do something with Astra if you assume a simplified model of some sorts, but at this point I don't really have exact suggestions on how to proceed. All the source code is located in this repository, but it's likely a non-trivial task to adapt it to a completely different model.

**Reply by askorikov** [reply]
PS: the code for calculating detector pixel vectors does not seem to be correct at the moment (that's why you see rays of different thicknesses in the reconstruction). Make sure you take into account the fact that sources are not located at [0, 0] 
coordinates.

**Reply by dannychancc** [reply]
Hi Askorikov,

Thanks for your help anyway! I will try to find the python code in this site.....

Danny
